isar-patches/0001-update-uboot-script-dynamic-load-the-dtbo-file.patch

```diff
@@ -0,0 +1,39 @@
+From 1df4154bbd69900c155bc2532214995a00826d6e Mon Sep 17 00:00:00 2001
+From: chao zeng <chao.zeng@siemens.com>
+Date: Mon, 10 Jan 2022 16:22:51 +0800
+Subject: [PATCH] update-uboot-script: dynamic load the dtbo file
+
+add the support to load the dtbo file dynamic
+Signed-off-by: chao zeng <chao.zeng@siemens.com>
```

Comment:
We already support overlays in u-boot-script. What is this doing here?

Reply:
currently , in the u-boot-script,we need to specify the `${OVERLAYS}` in the build stage.
but we need to set and load the dtbo dynamic at runtime. I think the existing overlay mechanism does not meet the needs

Reply:
U-Boot script is about implementing the distro boot interface. I doubt your extension is compliant here because it introduces a new generic variable ("name_overlay" is a TI invention, nothing generic). It also needlessly duplicates code in u-boot-script. We need a better fitting approach.

How about adding the needed logic as function to the default env of the IOT2050 boards and hooking that up in the board-specific configuration during image build (SCRIPT_PREPEND, wks file)?

But also keep secure boot in mind. We will then have to put all relevant artifacts for boot in the kernel into a FIT image or hard-code them into U-Boot. There will be no u-boot-script in its current form then.

Reply:
I think for secure boot, these M.2 functionality need be fixed down, then the dtbo is not needed and everything could be settle down on the build stage.

Reply:
True, there will be only insecurely modifiable U-Boot vars then. If switching the mode would be considered security-relevant (likely in most cases), this can't be controlled via U-Boot anymore.

Reply:
So here we need to add some security tuning for the building procedure, if the u-boot is built for secure boot then no dtbo will be used and a final dtb should be used directly.

This could be handled by the current secure boot development.

Reply:
For secure boot, we would likely have to enhance the image build recipes to patch the dtb embedded as `ti/k3-am6548-iot2050-advanced-m2.dtb` into the generated (and later signed) FIT image with the desired dtbos during build (`fdtoverlay`).

Reply:
> 
SCRIPT_PREPEND may could not implement this function.
if introduce the new var would break the generic flow. we would use fit image in the future,but now I think  we should discuss
more to make it  generic as isar lack of this mechanism

Reply:
Hmm, you need to apply the overlay after loading the fdt - obviously no "PREPEND" topic... Let me think... How about setting "OVERLAYS = "${my_overlays}"? That should allow to use existing logic with a U-Boot defined list of overlays.

Reply:
hmm. this change can not work as 
```
for OVERLAY in ${OVERLAYS}; do
....
done
```

doesn't to output the `${BOOT_CMD}`,

this change would not generate the boot.scr like this PR.it can not support load multiple dtbo

Reply:
I can't follow what you mean. Please give it a try (may require additional escaping of `$`) and show the difference in the script.

Reply:
Basically, adding `--sourceparams "overlays=\$\{name_overlays\}"` to the bootloader command in our wks file should do the trick.

Reply:
all this message should append to the BOOT_CMD, then mkimage generate the boot.scr. so we can execute the load logic in this scripts.

but already exists logic haven't write the 

```
for OVERLAY in ${OVERLAYS}; do
....
done
```


to the ${BOOT_CMD}. so that this logic is not correct

Reply:
OK, what you want to tell is that the look generates the script but is not embedded into the script. Then we need a patch that changes the generation loop into the runtime loop. But don't duplicate logic.

Reply:
Or, better, the existing loop body needs to detect if an entry in $OVERLAYS is a u-boot variable (starts with `$`) and then inject a loop into the script, rather than a single `fdt apply` command.

Reply:
OK. I will refine this patch.

Reply:
Untested first attempt (but build is running):
```
From 739fd15bec3b4e7f1235749ee2f6df136fc7e549 Mon Sep 17 00:00:00 2001
From: Jan Kiszka <jan.kiszka@siemens.com>
Date: Thu, 13 Jan 2022 09:20:23 +0100
Subject: [PATCH] u-boot-script: Allow OVERLAY to be a u-boot env variable as
 well

This allows to define the list of overlays also during boot-time,
retrieving it from a u-boot env variable. All we need to do for that is
adding a loop around loading and applying entries in the generated
script. That will interpret an entry during boot-time as list and
iterate over all entries.

Signed-off-by: Jan Kiszka <jan.kiszka@siemens.com>
---
 meta/recipes-bsp/u-boot-script/files/update-u-boot-script | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/meta/recipes-bsp/u-boot-script/files/update-u-boot-script b/meta/recipes-bsp/u-boot-script/files/update-u-boot-script
index eb2c2955..39b9fa0f 100755
--- a/meta/recipes-bsp/u-boot-script/files/update-u-boot-script
+++ b/meta/recipes-bsp/u-boot-script/files/update-u-boot-script
@@ -62,11 +62,13 @@ if [ -n "${OVERLAYS}" ]; then
 		if ! echo $OVERLAY | grep -q "^/"; then
 			OVERLAY_PATH=/usr/lib/linux-image-${KERNEL_VERSION}/
 		fi
-		echo "echo Loading ${OVERLAY_PATH}${OVERLAY}..." >> ${BOOT_CMD}
+		echo "for entry in ${OVERLAY}; do" >> ${BOOT_CMD}
+		echo "echo Loading ${OVERLAY_PATH}\${entry}..." >> ${BOOT_CMD}
 		echo "load \${devtype} \${devnum}:${ROOT_PARTITION}" \
-		     "\${overlay_addr_r} ${OVERLAY_PATH}${OVERLAY}" \
+		     "\${overlay_addr_r} ${OVERLAY_PATH}\${entry}" \
 		     >> ${BOOT_CMD}
 		echo "fdt apply \${overlay_addr_r}" >> ${BOOT_CMD}
+		echo "done" >> ${BOOT_CMD}
 	done
 fi
 
-- 
2.31.1
```

Reply:
Should work, but `setexpr` is not enabled in our U-Boot, and that needs to be changed first.

Also, my wks change was wrong:
```
diff --git a/wic/iot2050.wks.in b/wic/iot2050.wks.in
index 8ca915f9..acfb12fd 100644
--- a/wic/iot2050.wks.in
+++ b/wic/iot2050.wks.in
@@ -8,6 +8,6 @@
 # COPYING.MIT file in the top-level directory.
 #
 
-part / --source rootfs-u-boot --sourceparams "no_initrd=yes" --fstype ext4 --label rootfs --align 1024 --use-uuid
+part / --source rootfs-u-boot --sourceparams "no_initrd=yes" --fstype ext4 --label rootfs --align 1024 --use-uuid --sourceparams "overlays=\"\$\{name_overlays\}\""
 
 bootloader --ptable gpt --append "console=ttyS3,115200n8 earlycon=ns16550a,mmio32,0x02810000 mtdparts=47040000.spi.0:512k(ospi.tiboot3),2m(ospi.tispl),4m(ospi.u-boot),128k(ospi.env),128k(ospi.env.backup),1m(ospi.sysfw),64k(pru0-fw),64k(pru1-fw),64k(rtu0-fw),64k(rtu1-fw),-@8m(ospi.rootfs) rootwait ${EXTRA_KERNEL_PARAMS}"
 ```

Reply:
we do not use dtbo any more, so close this  session.